### PR TITLE
Replace typescript

### PR DESCRIPTION
**Pull request by segayuu**
One of the reasons why warehouse is hardly maintained is that it doesn't understand the type.
However, the warehouse package is too complex to be typed with DefinitelyTyped, and it is quite possible to make typing errors.

### CODE REVIEW

**Comment by YoshinoriN** [comment]
@segayuu 
IMHO we should release warehouse `v3.0` and update hexo's dependency `v3.0` before migrate to TypeScript.

**Comment by curbengh** [comment]
I just added warehouse@3 as part of hexo[ v4 milestone](https://github.com/hexojs/hexo/pull/3695).

**Comment by YoshinoriN** [comment]
@segayuu 
What point of view should we check? Just migrate js to ts?
For example, some instance member can specify `private`. Should we review such code?

**Comment by segayuu** [comment]
The purpose of this PR is to set `noImplicityAny` to true, and strict typing from that achievement is an arbitrary position. If the `private` attribute is added, it is necessary to modify the test file implementation.

**Comment by YuJianghao** [comment]
Do we have updates on this pr?

For now I think that hexo-warehouse is the best json db for tiny apps :p
Waiting for ts updates.

**Comment by tmirun** [comment]
I think the same @YuJianghao

**Comment by yoshinorin** [comment]
@hexojs/core 
I have been trying to migrate to TypeScript since last week.

I think maybe we can migrate it just a little bit more.
But, some tests have errors. I assume it causes by `rfdc` in `src/model.ts`.

> https://github.com/hexojs/warehouse/pull/60/files#diff-6bfea2ec3c1d21c90d6e1ecd704f380a6eeb44521e1da1a87b6a48ea691b6a43R2-R3

Dose someone would you help?

Thank you :)

**Comment by yoshinorin** [comment]
@hexojs/core 
Now all tests are pass.

File: test/scripts/database.ts [hunk 1]
```diff
@@ -1,17 +1,18 @@
-'use strict';
-
-const should = require('chai').should(); // eslint-disable-line
-const path = require('path');
-const Promise = require('bluebird');
-const sinon = require('sinon');
-const fs = Promise.promisifyAll(require('fs'));
+// @ts-nocheck
```
Comment (yoshinorin) [inline test/scripts/database.ts:1]: I add `@ts-nocheck` to all test files.
This workaround is temporarily.  I will update the test code.

**Comment by yoshinorin** [comment]
@hexojs/core
Now ready for review. Does someone would you please review?

I think we can improve our code more type safely and robustly. However, I think we need more time to do it. Also, I will fix if the current PR has a critical problem. But, IMHO I want to merge this PR once and improve our code step by step. 

Thank you :)

File: package.json [hunk 2]
```diff
@@ -24,27 +27,38 @@
...
     "mocha": "^10.0.0",
     "nyc": "^15.0.0",
-    "sinon": "^14.0.0"
+    "sinon": "^14.0.0",
+    "ts-node": "^10.8.1",
+    "typescript": "^4.7.4"
   },
   "engines": {
-    "node": ">=14"
+    "node": ">=14 <18.5"
```
Comment (yoshinorin) [inline package.json:62]: This is a workaround.
The `Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close` error occurs with Node `18.6`. I don't know this problem is a bug or a spec change.

Comment (yoshinorin) [reply package.json:62]: IMHO we should delete `<18.5` before release the new version (`5.0.0`)

Comment (yoshinorin) [reply package.json]: Deleted

**Comment by yoshinorin** [comment]
Updated: Merge the `master` branch to this PR branch.

File: src/database.ts [hunk 3]
```diff
@@ -69,7 +69,17 @@ async function exportAsync(database, path) {
   }
 }
 
+type DatabaseOptions = {
+  version: number,
+  path: string,
+  onUpgrade: (...args: any[]) => any,
+  onDowngrade: (...args: any[]) => any
+};
```
Comment (SukkaW) [inline src/database.ts:77]:
```suggestion
interface DatabaseOptions {
  version: number,
  path: string,
  onUpgrade: (...args: any[]) => any,
  onDowngrade: (...args: any[]) => any
}
```

Prefer `interface` when possible.

**Review by SukkaW — CHANGES_REQUESTED** [review]

File: src/lib/jsonstream/index.ts [hunk 4]
```diff
@@ -1,7 +1,5 @@
-'use strict';
-
-const through2 = require('through2');
-const Parser = require('jsonparse');
+import through2 from 'through2';
```
Comment (SukkaW) [inline src/lib/jsonstream/index.ts:1]: IIRC `through2` has its type definition `@types/through2`?

File: src/database.ts (hunk 3 above)
Comment (yoshinorin) [reply src/database.ts]: Done :)

File: src/lib/jsonstream/index.ts (hunk 4 above)
Comment (yoshinorin) [reply src/lib/jsonstream/index.ts:1]: Done :)

**Review by renbaoshuo — APPROVED** [review]

**Review by stevenjoezhang — APPROVED** [review]

**Pull request merged**